src/lib/expression/expression_utils.cpp

```diff
@@ -233,22 +233,17 @@ bool expression_contains_placeholders(const std::shared_ptr<AbstractExpression>&
   auto placeholder_found = false;
 
   visit_expression(expression, [&](const auto& sub_expression) {
-    const auto parameter_expression = std::dynamic_pointer_cast<ParameterExpression>(sub_expression);
-    if (parameter_expression) {
-      placeholder_found |= parameter_expression->parameter_expression_type == ParameterExpressionType::ValuePlaceholder;
-    }
-
+    placeholder_found |= std::dynamic_pointer_cast<PlaceholderExpression>(sub_expression) != nullptr;
     return ExpressionVisitation::VisitArguments;
   });
 
   return placeholder_found;
 }
 
 std::optional<AllTypeVariant> expression_get_value_or_parameter(const AbstractExpression& expression) {
-  if (expression.type == ExpressionType::Parameter) {
-    const auto& parameter_expression = static_cast<const ParameterExpression&>(expression);
-    DebugAssert(parameter_expression.value(), "ParameterExpression doesn't have a value set");
-    return *parameter_expression.value();
+  if (const auto* correlated_parameter_expression = dynamic_cast<const CorrelatedParameterExpression*>(&expression)) {
```

Comment:
Why do we use raw pointers here?

Reply:
Because  "expression" is a reference, and rightfully so. Fundamentally, what is your problem with this exact pattern?

Reply:
I think it's great for functions to take values by reference if they do not need to store the object. It's less redundant typing of "std::shared_ptr<>" as the objects gets passed down the callstack and therefore increases readability. The promise that this object will not get stored persistently is right there in the function sig, and that's also good. 

Now if we, somewhere down the callstack, want to use the RTTI of this object, what do we do? Change the signature of all functions above or use the tool that c++ gives us: dynamic_cast?

Reply:
My worry is that once we start taking addresses of things and throwing around raw pointers, the dam is broken. While you know when to use raw pointers and when not, people with less C++ experience will not.

In isolation, I fully agree with you on this one. On a bigger scale, I feel that, in our circumstances, the project benefits from a general ban on raw pointers. This would be different for a project with a more stable group of developers who can be expected to know the difference of a raw pointer, a reference, a smart pointer, and a reference to a smart pointer.

Reply:
What is a pointer?

Reply:
I and others (mostly me, though ;) ) managed to sneak in 30 dynamic_casts into libhyrise, so the question presents itself whether this PR is the right place to start enforcing a new policy.

That being said, I absolutely agree with you that certain strict rules, even if they might be cumbersome to work around, can help prevent common bugs.

But maybe "Don't use raw points, never! never!" is too strict of a rule. Maybe "Don't store raw pointers persistently in an object" is more practical. Maybe accompanied by "Use dynamic_cast only to access RTTI, along the lines of `if (auto* a = dynamic_cast<>()) { /* Do something with a*/}`, prefer to use `dynamic_pointer_cast` wherever you can."